### ISSUE TITLE
#persisted? return true if object saved in nested transaction(with requires_new) and rollback in parent transaction.

### ISSUE THREAD

**Issue by eroku** [rails/rails#29320]
### Steps to reproduce

Although there is nested transaction, but nested transaction have not been noticed it rollback by `RealTransaction`.

```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  gem "arel", github: "rails/arel"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_persisted_in_nested_transaction_with_requires_new
    post_one = Post.new
    post_two = Post.new

    Post.transaction do
      post_one.save
      
      Post.transaction(requires_new: true) do
        post_two.save
      end

      assert_equal true, post_one.persisted?
      assert_equal true, post_two.persisted?

      raise ActiveRecord::Rollback
    end

    assert_equal 0, Post.count
    assert_equal false, post_one.persisted?
    assert_equal false, post_two.persisted?
  end
end
```

### Expected behavior
rollback transaction, so `post_two.persisted?` should be return `false`.

### Actual behavior
```
  1) Failure:
BugTest#test_persisted_in_nested_transaction_with_requires_new [test.rb:60]:
Expected: false
  Actual: true
```

### System configuration
**Rails version**:
Rails 5.1.1

**Ruby version**:
ruby 2.3.4p301

**Comment by eileencodes** [comment]
I see what you're seeing but I'm not sure if there's a way to fix it easily. This has been an issue since at least 4.2 (I tested 4-2-stable, 5-0-stable and master)

When `persisted?` is called Active Record checks the `sync_with_transaction_state` which updates the state if necessary. When the first transaction is opened it is actually rolled back and gets a state of `rolledback`. The nested transaction is actually a `SAVEPOINT` but when that transaction is closed it doesn't know yet that it will be rolledback so it gets that state of `committed`.

The outside transaction then gets the `rolledback` state so the `persisted?` check on the first record is correctly marked as `false` but the nested transaction is still marked as `committed` since it wasn't actually rolled back.

```
D, [2017-06-02T15:46:37.392390 #91604] DEBUG -- :    (0.0ms)  begin transaction
D, [2017-06-02T15:46:37.393033 #91604] DEBUG -- :   SQL (0.1ms)  INSERT INTO "posts" DEFAULT VALUES
D, [2017-06-02T15:46:37.393240 #91604] DEBUG -- :    (0.0ms)  SAVEPOINT active_record_1
D, [2017-06-02T15:46:37.393507 #91604] DEBUG -- :   SQL (0.0ms)  INSERT INTO "posts" DEFAULT VALUES
D, [2017-06-02T15:46:37.393652 #91604] DEBUG -- :    (0.0ms)  RELEASE SAVEPOINT active_record_1
TRANSACTION_STATE: 70229796152500 => committed
D, [2017-06-02T15:46:37.393812 #91604] DEBUG -- :    (0.0ms)  rollback transaction
D, [2017-06-02T15:46:37.394203 #91604] DEBUG -- :    (0.1ms)  SELECT COUNT(*) FROM "posts"
TRANSACTION_STATE: 70229796175020 => rolledback
TRANSACTION_STATE: 70229796152500 => committed
```

While I do think this is unexpected I'm not sure if there's an easy fix. This would require marking the inside transaction as not `committed` after the first transaction completed. When it was originally marked as `committed` it was in fact `committed` and that would require retroactively changing the status of a transaction. Post 2 gets rolled back because everything that happened inside transaction 1 is rolled back, but transaction 2 was still committed.

**Comment by matthewd** [comment]
Could the nested transaction adopt its parent's status upon "commit", instead of over-eagerly declaring itself committed? That sounds arguably closer to the semantics of `RELEASE SAVEPOINT`, at least.

**Comment by eileencodes** [comment]
Ok so if we take the commit during the savepoint and mark it as the parent status (at this point nil because the parent transaction hasn't completed yet) it still is marked as `persisted` in the db. I was wrong about it being directly related to the transaction status.

Looking closer `post_one` is marked as a `@new_record` because it's never successfully saved to the database. `post_two` is not marked as a `@new_record` because it was saved.

I think to actually fix this we'd need to figure out how to mark the previous record state as new. At this point we don't know there was a parent transaction (because the nested transaction finished) to rely on to change the state of the record. We don't know that the parent transaction is going to fail so our only choice is to mark the record as persisted.

I don't know if parent transaction is aware there is a second record that was committed to be able to change it's state during the rollback.

**Comment by eileencodes** [comment]
Hey @eroku we fixed this in ef48089fd9e86a708cb2215b4f1bb33444816c88. Thanks for reporting! It does change long standing behavior so I'm not going to backport it.

**Comment by mdworken** [comment]
Hi @eileencodes! This is my first foray into the core rails project, so definitely feel free to DM me if you have any feedback or suggestions :) 

----

Can we consider re-opening this issue? I'm consistently seeing it reproduce by adding the file below to a number of my existing (>=5.2) projects and running `bundle exec rails r <file_name>`. It's the same file as above but I removed the Gemfile definition and the sqlite3 memory database connection in favor of what my app was already using. This test is confirmed failing when run within https://github.com/rubyforgood/abalone, and fails in the same spot as the one it was modified from above.

**### System configuration**
ruby '2.6.3'
rails'  5.2.3'
pg  '1.2.3

`require "active_record"
require "minitest/autorun"
require "logger"
ActiveRecord::Base.establish_connection
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
  create_table :comments, force: true do |t|
    t.integer :post_id
  end
end
class Post < ActiveRecord::Base
  has_many :comments
end
class Comment < ActiveRecord::Base
  belongs_to :post
end
class BugTest < Minitest::Test
  def test_persisted_in_nested_transaction_with_requires_new
    post_one = Post.new
    post_two = Post.new
Post.transaction do
      post_one.save
      Post.transaction(requires_new: true) do
        post_two.save
      end
      assert_equal true, post_one.persisted?
      assert_equal true, post_two.persisted?
      raise ActiveRecord::Rollback
    end
    assert_equal 0, Post.count
    assert_equal false, post_one.persisted?
    assert_equal false, post_two.persisted?
  end
end`

**Comment by eileencodes** [comment]
We don't usually reopen 3-year old issues especially since a lot has changed, please open a new issue with the provided reproduction script. If this is the same issue you would have seen it on 4.2, 5.0, and 5.1 as well since this was only fixed in 5.2 and has probably always been an issue. If you weren't seeing it on 4.2-5.2 then it's likely a different issue. 

Opening a new issue is also better because at the moment I won't have time to dig into this and a new issue will be more likely get attention from another community member.

**Comment by mdworken** [comment]
@eileencodes got it, thanks! Wasn't sure of the guidance on new vs reopen so appreciate the clarity :)
